code/modules/food_and_drinks/kitchen_machinery/food_cart.dm

```diff
@@ -10,151 +6,259 @@
-	for(var/datum/reagent/R in mixer.reagents.reagent_list)
-		dat += "[R.name]: [R.volume] "
-		dat += "<a href='?src=[REF(src)];transfer=[R.type]'>Transfer back</a>"
-		if (glasses > 0)
-			dat += "<a href='?src=[REF(src)];m_pour=[R.type]'>Pour in a glass</a>"
-		dat += "<br>"
-	dat += "</div><br><b>STORED FOOD</b><br><div class='statusDisplay'>"
-	for(var/V in stored_food)
-		if(stored_food[V] > 0)
-			dat += "<b>[V]: [stored_food[V]]</b> <a href='?src=[REF(src)];dispense=[V]'>Dispense</a><br>"
-	dat += "</div><br><a href='?src=[REF(src)];refresh=1'>Refresh</a> <a href='?src=[REF(src)];close=1'>Close</a>"
-
-	var/datum/browser/popup = new(user, "foodcart","Food Cart", 500, 350, src)
-	popup.set_content(dat)
-	popup.open()
-
-/obj/machinery/food_cart/proc/isFull()
-	return food_stored >= STORAGE_CAPACITY
-
-/obj/machinery/food_cart/attackby(obj/item/O, mob/user, params)
-	if(O.tool_behaviour == TOOL_WRENCH)
-		default_unfasten_wrench(user, O, 0)
-		return TRUE
-	if(istype(O, /obj/item/reagent_containers/food/drinks/drinkingglass))
-		var/obj/item/reagent_containers/food/drinks/drinkingglass/DG = O
-		if(!DG.reagents.total_volume) //glass is empty
-			qdel(DG)
-			glasses++
-			to_chat(user, span_notice("[src] accepts the drinking glass, sterilizing it."))
-	else if(istype(O, /obj/item/reagent_containers/food/snacks))
-		if(isFull())
-			to_chat(user, span_warning("[src] is at full capacity."))
-		else
-			var/obj/item/reagent_containers/food/snacks/S = O
-			if(!user.transferItemToLoc(S, src))
-				return
-			if(stored_food[sanitize(S.name)])
-				stored_food[sanitize(S.name)]++
-			else
-				stored_food[sanitize(S.name)] = 1
-	else if(istype(O, /obj/item/stack/sheet/glass))
-		var/obj/item/stack/sheet/glass/G = O
-		if(G.get_amount() >= 1)
-			G.use(1)
-			glasses += 4
-			to_chat(user, span_notice("[src] accepts a sheet of glass."))
-	else if(istype(O, /obj/item/storage/bag/tray))
-		var/obj/item/storage/bag/tray/T = O
-		for(var/obj/item/reagent_containers/food/snacks/S in T.contents)
-			if(isFull())
-				to_chat(user, span_warning("[src] is at full capacity."))
-				break
-			else
-				if(SEND_SIGNAL(T, COMSIG_TRY_STORAGE_TAKE, S, src))
-					if(stored_food[sanitize(S.name)])
-						stored_food[sanitize(S.name)]++
-					else
-						stored_food[sanitize(S.name)] = 1
-	else if(O.is_drainable())
+//For adding items and reagents to storage
+/obj/machinery/food_cart/attackby(obj/item/A, mob/user, params)
+	//Depending on the item, either attempt to store it or ignore it
+	if(istype(A, /obj/item/reagent_containers/food/snacks))
+		storage_single(A)
+		return
+	else if(istype(A, /obj/item/reagent_containers/food/drinks/drinkingglass))
+		//Check if glass is empty
+		if(!A.reagents.total_volume)
+			//Delete glass and increment glass_quantity by 1
+			qdel(A)
+			glass_quantity++
+			user.visible_message(span_notice("[user] inserts [A] into [src]."), span_notice("You insert [A] into [src]."))
```

Comment:
move the `qdel(A)` to after the visible_message call that references `A` to avoid use-after-free